File: polo_client/example/polo_client_example.dart
import 'package:polo_client/polo_client.dart';

class UserType implements PoloType {
  final String name;
  final int age;

  UserType({required this.name, required this.age});

  @override
  factory UserType.fromMap(Map<String, dynamic> map) {
    return UserType(name: map['name'], age: map['age']);
  }

  @override
  Map<String, dynamic> toMap() {
    return {'name': name, 'age': age};
  }
}

void main() async {
  // Polo Client
  PoloClient client = await Polo.connect("ws://127.0.0.1:3000/");

  client.registerType(
    PoloTypeAdapter<UserType>(
      toMap: ((type) => type.toMap()),
      fromMap: (map) => UserType.fromMap(map),
    ),
  );

  client.onConnect(() {
    print("Client Connected to Server");

    client.send('dynamic', "Ayush");
    client.send('dynamic', 1);
    client.send('dynamic', 3.14);
    client.send('dynamic', true);
    client.send('dynamic', [1, 2, 3]);
    client.send('dynamic', null);
    client.send('dynamic', {
      "String": {"dynamic": true}
    });
    client.send<String>('message', "Hello from Client");
    client.send<UserType>('userJoined', UserType(name: "Ayush", age: 22));
  });

  client.onEvent('dynamic', (dyn) {
    print("Dynamic: $dyn : ${dyn.runtimeType}");
  });

  client.onEvent<String>('message', (message) {
    print("Message: $message : ${message.runtimeType}");
  });

  client.onEvent<UserType>('userJoined', (user) {
    print("userJoined : ${user.toMap()} : ${user.runtimeType}");
  });

  client.onDisconnect(() {
    print("Client Disconnected from Server");
  });

  client.listen();
}

File: polo_client_ts/example/polo_client_example.ts
import { Polo, PoloClient } from "../polo_client.ts";

interface UserType {
  name: string;
  age: number;
}

// Polo Client
const client: PoloClient = await Polo.connect(
  "ws://127.0.0.1:3000/",
);

client.onConnect(() => {
  console.log("Client Connected to Server");

  setInterval(() => {
    client.send<string>("polo:ping", new Date().toISOString());
  }, 1000);

  client.send("dynamic", "Ayush");
  client.send("dynamic", 1);
  client.send("dynamic", 3.14);
  client.send("dynamic", true);
  client.send("dynamic", [1, 2, 3]);
  client.send("dynamic", null);
  client.send("dynamic", {
    "String": { "dynamic": true },
  });
  client.send<string>("message", "Hello from Client");
  client.send<UserType>("userJoined", <UserType> { name: "Ayush", age: 22 });
});

client.onEvent<string>("polo:pong", (dateTime) => {
  console.log(
    `Ping: ${((new Date().getTime()) - (new Date(dateTime).getTime()))} ms`,
  );
});

client.onEvent("dynamic", (dyn) => {
  console.log(`Dynamic: ${dyn} : ${typeof dyn}`);
});

client.onEvent<string>("message", (message) => {
  console.log(`Message: ${message} : ${typeof message}`);
});

client.onEvent<UserType>("userJoined", (user) => {
  console.log(`userJoined : ${JSON.stringify(user)} : ${typeof user}`);
});

client.onDisconnect(() => {
  console.log("Client Disconnected from Server");
});

client.listen();
